### DISCUSSION TITLE
Analysis of templ AST for tooling and framework development

### DISCUSSION THREAD

**Discussion by tylermmorton** [a-h/templ#586] [General]
The ability to inspect (and modify) component AST directly is a huge boon for tooling developers in the JSX/React world. This enables things like `prettier` and `eslint` to format and lint code effectively. It also opens the door for code migration via AST (dubbed codemods), where transformations are run against UI codebases automatically.

At another level, while building a MVC system, it may be helpful for the controller to inspect the view layer for certain properties. A great example of this is Remix's 'nested routes', where an `<Outlet/>` element is added in the view layer, and the controller layer then knows to look for that tag and handle the nesting logic. 

I wanted to bring this up as a topic of discussion, because to me, great support for static analysis is vital to so many parts of web development. FWIW `html/template`s have [ParseTree](https://github.com/tylermmorton/tmpl/blob/develop/traverse.go)s that enable this type of work. 

I will have to dig around in `a-h/templ/parse` package to see what is available. Is there a public API that is meant for this use case and are there any examples of tools being developed with it?

Also I know that `templ` files are compiled at build time and converted into Go functions so all of the templ AST is unavailable during runtime. I'm wondering if there is a runtime variant of the compiler that could be used in situations where server frameworks need to inspect template structure?

Thanks in advance for the discussion!

**Comment by joerdav** [comment]
I agree that this would be great for developing external tools, such as stricter formatters with rules that are too opinionated to be the defaults, and even security checkers. 

I do believe that the current parser package should be able to support this use case though I've not seen any examples yet. There could be some quality of life changes that would make this easier such as a `Walk` method to iterate nodes.

**Comment by dgrundel** [comment]
Apologies for resurrecting a fairly old thread. My team is using Templ in a large project with lots of contributors and we're creating a Templ linter to help enforce some of our own coding standards. 

Part of creating a good linter is being able to point to a specific source location where problem code exists. However, this is a challenge using the existing AST nodes. Many of them do not include a `Range`, while others only provide a `Range` for part of the source (e.g. `Element` provides `NameRange`).

@joerdav I did also have to implement my own `Walk` function. :D 

Any opposition to a pull request that would add additional position information to existing nodes? Have a preference as to using `Range`s or simple indexes?

**Reply by AdamVig** [reply]
@dgrundel I'm trying to lint Templ at my company too, but have added partial support to @html-eslint/eslint-plugin rather than using the official AST. Is there anything you can share about your Templ linter, even just the architecture of it, and how it has been working out for your team? My email is on my profile if you're not able to share publicly. Thanks!

**Reply by dgrundel** [reply]
Hey Adam, 

I'll do my best to sketch an outline here for you.

The basic premise is that you end up with a CLI tool you can run with a syntax like this:

```
templ-lint --config=some_config.yaml /path/to/files/**/*.templ
```

The high-level components are:
- Rules
- Runner
- TreeWalker

## Rules
My rule interface is basically:

```go
type Rule interface {
  Name() string
  Visitor() Visitor
}
```

Where `Visitor` looks and works _mostly_ like the [parser's built-in interface](https://github.com/a-h/templ/blob/main/parser/v2/visitor.go). The difference being that my visitor funcs get an extra context object:

```
type LinterVisitor[N] = func (node N, nodeCtx NodeContext) error
```

`NodeContext` let's me provide the visitor with a little extra contextual information about the node, the most interesting being a reference to its **chain of parent nodes**. This ancestry info is important when you want to write a rule like "make sure all `<img>` tags have an `alt` attribute", since those attributes are just nodes in the AST.

In the current implementation, the visitor simply returns an `error` to indicate a node has failed a check. There is an **error decorator** that exists elsewhere that adds extra info to the error for output purposes, including the name of the rule that generated the error, and the location of the error (file, line, column).

## Runner
The runner is the main entrypoint and orchestrator of the linting process. 
- parses the CLI args (I just used the standard `flags` library)
- generates a set of files to run the rules against (I used `doublestar` for the glob resolution)
- parses the config file to figure out what rules to run
- runs the rules (more on this below)
- decorates the errors returned by the rules
- emits output to stdout/stderr

In reality you probably don't want all of this to be in a single component. But from a general architecture perspective, you need these things to happen and be orchestrated by something like this.

## Running the rules

To actually execute rules on files, I created what I call a `TreeWalker`. The implementation looks a lot like [the visitor implementation](https://github.com/a-h/templ/blob/7be7dd66fc7ff652d4ac8e5ee2f222442b73277f/parser/v2/visitor/visitor.go) in the parser...probably because I copied-and-pasted it into my code as a starting point. :)

`TreeWalker` takes as inputs a `Visitor` (as previously seen in the Rule interface) and a `TemplateFile` as returned by `parser.Parse` ([this thing](https://github.com/a-h/templ/blob/main/parser/v2/templatefile.go#L13)). It recursively walks the AST and invokes the rule's provided `Visitor` on every node. It also keeps track of the hierarchy of the AST and provides that helpful `NodeContext` object to all the invoked visitor funcs.

When rules return an `error`, they are passed to an error decorator (along with the `NodeContext`!) which wraps them with a richer interface, something like:

```go
type LintError interface {
  Error() string
  RuleName() string
  FilePath() string
  Line() int
  Col() int
}
```

Basically everything you need to make a nice, eslint-esque rule output.

---

Aaaaaanyway, this is getting pretty long. I think I covered the basics, I hope this is helpful. Happy to answer follow-up questions. This is all from memory so if something seems off it's probably not you. :)

**Reply by AdamVig** [reply]
Thank you so much for the detailed response! This seems great. I'm taking a slightly different approach: I built a super simple Go program that runs the Templ parser and prints the AST as JSON to stdout.
```go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"os"

	"github.com/a-h/templ/parser/v2"
)

func main() {
	inputBytes, err := io.ReadAll(os.Stdin)
	if err != nil {
		fmt.Fprintln(os.Stderr, "Error reading input:", err)
		return
	}
	input := string(inputBytes)
	templateFile, err := parser.ParseString(input)
	if err != nil {
		fmt.Fprintln(os.Stderr, "Error parsing template:", err)
		return
	}

	nodesJSON, err := json.MarshalIndent(templateFile.Nodes, "", "  ")
	if err != nil {
		fmt.Fprintln(os.Stderr, "Error marshaling nodes to JSON:", err)
		return
	}
	fmt.Println(string(nodesJSON))
}
```
Then I built an ESLint plugin that invokes the binary and converts the AST to the one expected by [HTML ESLint](https://github.com/yeonjuan/html-eslint) so that all the rules from that plugin can run on Templ files. It needs some straightening out, partially through enhancing the Templ AST, but I think the approach will work well and can be expanded to allow ESLint to run on CSS and JavaScript in Templ files as well.

Anyway, just figured I'd share!

**Reply by dgrundel** [reply]
Nice! That's very clever. I'll share that with my team as well!

**Comment by AdamVig** [comment]
I want to share that my ESLint plugin for Templ is now open source: https://github.com/AdamVig/eslint-plugin-templ
We have been using it at work for a couple of months with no problems.

cc:
@dgrundel Thanks for kicking off the AST enhancement effort!
@whalenrp I saw #1402 and thought you might be interested!
@a-h Thanks for patiently reviewing all of my pull requests! I've got one more coming and then I think I can switch the plugin off of my Templ fork.
@joerdav For visibility
